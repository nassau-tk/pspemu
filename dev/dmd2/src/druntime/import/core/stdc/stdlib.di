// D import file generated from 'src\core\stdc\stdlib.d'
module core.stdc.stdlib;
private import core.stdc.config;

public import core.stdc.stddef;

extern (C) nothrow 
{
    struct div_t
{
    int quot;
    int rem;
}
    struct ldiv_t
{
    int quot;
    int rem;
}
    struct lldiv_t
{
    long quot;
    long rem;
}
    enum EXIT_SUCCESS = 0;
    enum EXIT_FAILURE = 1;
    enum MB_CUR_MAX = 1;
    version (Windows)
{
    enum RAND_MAX = 32767;
}
else
{
    version (linux)
{
    enum RAND_MAX = 2147483647;
}
else
{
    version (OSX)
{
    enum RAND_MAX = 2147483647;
}
else
{
    version (FreeBSD)
{
    enum RAND_MAX = 2147483647;
}
else
{
    version (Solaris)
{
    enum RAND_MAX = 32767;
}
else
{
    static assert(false,"Unsupported platform");
}
}
}
}
}
    double atof(in char* nptr);
    int atoi(in char* nptr);
    c_long atol(in char* nptr);
    long atoll(in char* nptr);
    double strtod(in char* nptr, char** endptr);
    float strtof(in char* nptr, char** endptr);
    real strtold(in char* nptr, char** endptr);
    c_long strtol(in char* nptr, char** endptr, int base);
    long strtoll(in char* nptr, char** endptr, int base);
    c_ulong strtoul(in char* nptr, char** endptr, int base);
    ulong strtoull(in char* nptr, char** endptr, int base);
    int rand();
    void srand(uint seed);
    void* malloc(size_t size);
    void* calloc(size_t nmemb, size_t size);
    void* realloc(void* ptr, size_t size);
    void free(void* ptr);
    void abort();
    void exit(int status);
    int atexit(void function() func);
    void _Exit(int status);
    char* getenv(in char* name);
    int system(in char* string);
    void* bsearch(in void* key, in void* base, size_t nmemb, size_t size, int function(in void*, in void*) compar);
    void qsort(void* base, size_t nmemb, size_t size, int function(in void*, in void*) compar);
    pure int abs(int j);

    pure c_long labs(c_long j);

    pure long llabs(long j);

    div_t div(int numer, int denom);
    ldiv_t ldiv(c_long numer, c_long denom);
    lldiv_t lldiv(long numer, long denom);
    int mblen(in char* s, size_t n);
    int mbtowc(wchar_t* pwc, in char* s, size_t n);
    int wctomb(char* s, wchar_t wc);
    size_t mbstowcs(wchar_t* pwcs, in char* s, size_t n);
    size_t wcstombs(char* s, in wchar_t* pwcs, size_t n);
    version (DigitalMars)
{
    void* alloca(size_t size);
}
}

