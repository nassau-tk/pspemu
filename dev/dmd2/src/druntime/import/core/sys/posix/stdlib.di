// D import file generated from 'src\core\sys\posix\stdlib.d'
module core.sys.posix.stdlib;
private import core.sys.posix.config;

public import core.stdc.stdlib;

public import core.sys.posix.sys.wait;

extern (C) 
{
    version (linux)
{
    int posix_memalign(void**, size_t, size_t);
}
else
{
    version (FreeBSD)
{
    int posix_memalign(void**, size_t, size_t);
}
}
    version (linux)
{
    int setenv(in char*, in char*, int);
    int unsetenv(in char*);
    void* valloc(size_t);
}
else
{
    version (OSX)
{
    int setenv(in char*, in char*, int);
    int unsetenv(in char*);
    void* valloc(size_t);
}
else
{
    version (FreeBSD)
{
    int setenv(in char*, in char*, int);
    int unsetenv(in char*);
    void* valloc(size_t);
}
}
}
    version (linux)
{
    int rand_r(uint*);
}
else
{
    version (OSX)
{
    int rand_r(uint*);
}
else
{
    version (FreeBSD)
{
    int rand_r(uint*);
}
}
}
    version (linux)
{
    c_long a64l(in char*);
    double drand48();
    char* ecvt(double, int, int*, int*);
    double erand48(ref ushort[3]);
    char* fcvt(double, int, int*, int*);
    char* gcvt(double, int, char*);
    int getsubopt(char**, in char**, char**);
    int grantpt(int);
    char* initstate(uint, char*, size_t);
    c_long jrand48(ref ushort[3]);
    char* l64a(c_long);
    void lcong48(ref ushort[7]);
    c_long lrand48();
    char* mktemp(char*);
    c_long mrand48();
    c_long nrand48(ref ushort[3]);
    int posix_openpt(int);
    char* ptsname(int);
    int putenv(char*);
    c_long random();
    char* realpath(in char*, char*);
    ushort seed48(ref ushort[3]);
    void setkey(in char*);
    char* setstate(in char*);
    void srand48(c_long);
    void srandom(uint);
    int unlockpt(int);
    static if(__USE_LARGEFILE64)
{
    int mkstemp64(char*);
    alias mkstemp64 mkstemp;
}
else
{
    int mkstemp(char*);
}
}
else
{
    version (OSX)
{
    c_long a64l(in char*);
    double drand48();
    char* ecvt(double, int, int*, int*);
    double erand48(ref ushort[3]);
    char* fcvt(double, int, int*, int*);
    char* gcvt(double, int, char*);
    int getsubopt(char**, in char**, char**);
    int grantpt(int);
    char* initstate(uint, char*, size_t);
    c_long jrand48(ref ushort[3]);
    char* l64a(c_long);
    void lcong48(ref ushort[7]);
    c_long lrand48();
    char* mktemp(char*);
    int mkstemp(char*);
    c_long mrand48();
    c_long nrand48(ref ushort[3]);
    int posix_openpt(int);
    char* ptsname(int);
    int putenv(char*);
    c_long random();
    char* realpath(in char*, char*);
    ushort seed48(ref ushort[3]);
    void setkey(in char*);
    char* setstate(in char*);
    void srand48(c_long);
    void srandom(uint);
    int unlockpt(int);
}
else
{
    version (FreeBSD)
{
    c_long a64l(in char*);
    double drand48();
    double erand48(ref ushort[3]);
    int getsubopt(char**, in char**, char**);
    int grantpt(int);
    char* initstate(uint, char*, size_t);
    c_long jrand48(ref ushort[3]);
    char* l64a(c_long);
    void lcong48(ref ushort[7]);
    c_long lrand48();
    char* mktemp(char*);
    int mkstemp(char*);
    c_long mrand48();
    c_long nrand48(ref ushort[3]);
    int posix_openpt(int);
    char* ptsname(int);
    int putenv(char*);
    c_long random();
    char* realpath(in char*, char*);
    ushort seed48(ref ushort[3]);
    void setkey(in char*);
    char* setstate(in char*);
    void srand48(c_long);
    void srandom(uint);
    int unlockpt(int);
}
}
}
}
