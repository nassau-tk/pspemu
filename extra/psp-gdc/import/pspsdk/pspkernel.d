module pspsdk.pspkernel;

/*
#include <pspuser.h>
#include <pspiofilemgr_kernel.h>
#include <psploadcore.h>
#include <pspstdio_kernel.h>
#include <pspsysreg.h>
#include <pspkdebug.h>
#include <pspintrman_kernel.h>
#include <pspmodulemgr_kernel.h>
*/

/**
 * Set the $pc register to a kernel memory address.
 *
 * When the PSP's kernel library stubs are called, they expect to be accessed
 * from the kernel's address space.  Use this function to set $pc to the kernel
 * address space, before calling a kernel library stub.
 */
/*
#define pspKernelSetKernelPC()  \
{     \
	__asm__ volatile (      \
	"la     $8, 1f\n\t"     \
	"lui    $9, 0x8000\n\t" \
	"or     $8, $9\n\t"     \
	"jr     $8\n\t"         \
	" nop\n\t"              \
	"1:\n\t"                \
	: : : "$8", "$9");      \
	sceKernelIcacheClearAll(); \
}
*/

/**
 * Set the $pc register to a user memory address.
 *
 */
/*
#define pspKernelSetUserPC()  \
{     \
	__asm__ volatile (      \
	"la     $8, 1f\n\t"     \
	"li     $9, 0x7FFFFFFF\n\t" \
	"and    $8, $9\n\t"     \
	"jr     $8\n\t"         \
	" nop\n\t"              \
	"1:\n\t"                \
	: : : "$8", "$9");      \
	sceKernelIcacheClearAll(); \
}
*/