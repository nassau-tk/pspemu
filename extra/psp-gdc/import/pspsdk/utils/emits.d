module pspsdk.utils.emits;

import std.string;

extern (C):

void emitInt    (int   v);
void emitFloat  (float v);
void emitString (char *v);
void emitComment(char *v);
void emitMemoryBlock(void *address, uint size);
void emitHex(void *address, uint size);
void emitDString(char[] v) { emitString(toStringz(v)); }