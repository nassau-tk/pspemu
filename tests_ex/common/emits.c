void emitInt(int v) {
	asm("syscall 0x1010");
}

void emitFloat(float v) {
	asm("syscall 0x1011");
}

void emitString(char *v) {
	asm("syscall 0x1012");
}

void emitMemoryBlock(void *address, unsigned int size) {
	asm("syscall 0x1013");
}
