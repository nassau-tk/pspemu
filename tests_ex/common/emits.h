void emitInt(int v);
void emitFloat(float v);
void emitString(char *v);
void emitMemoryBlock(void *address, unsigned int size);
