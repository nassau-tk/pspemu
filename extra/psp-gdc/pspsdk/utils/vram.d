module pspsdk.utils.vram;

import pspsdk.pspge;
import pspsdk.pspgu;

static uint staticOffset = 0;

static uint getMemorySize(uint width, uint height, uint psm) {
	switch (psm) {
		case GU_PSM_T4: return (width * height) >> 1;
		case GU_PSM_T8: return width * height;
		case GU_PSM_5650, GU_PSM_5551, GU_PSM_4444, GU_PSM_T16: return 2 * width * height;
		case GU_PSM_8888, GU_PSM_T32: return 4 * width * height;
		default: return 0;
	}
}

void* getStaticVramBuffer(uint width, uint height, uint psm) {
	uint memSize = getMemorySize(width,height,psm);
	void* result = cast(void*)staticOffset;
	staticOffset += memSize;
	return result;
}

void* getStaticVramTexture(uint width, uint height, uint psm) {
	void* result = getStaticVramBuffer(width,height,psm);
	return cast(void*)((cast(uint)result) + (cast(uint)sceGeEdramGetAddr()));
}
