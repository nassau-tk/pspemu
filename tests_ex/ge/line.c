#pragma compile, "%PSPSDK%/bin/psp-gcc" -I. -I"%PSPSDK%/psp/sdk/include" -L. -L"%PSPSDK%/psp/sdk/lib" -D_PSP_FW_VERSION=150 -Wall -g line.c ../common/emits.c ../common/vram.c -lpspgum -lpspgu -lm -lpsprtc -lpspdebug -lpspdisplay -lpspge -lpspsdk -lc -lpspuser -lpspkernel -o line.elf
#pragma compile, "%PSPSDK%/bin/psp-fixup-imports" line.elf

#include <pspkernel.h>
#include <pspdisplay.h>
#include <pspdebug.h>
#include <stdlib.h>
#include <pspctrl.h>
#include <pspgu.h>
#include <pspgum.h>

#include "../common/vram.h"
#include "../common/emits.h"

PSP_MODULE_INFO("Test Line Drawing", 0, 1, 1);

static unsigned int __attribute__((aligned(16))) list[262144];

struct Vertex { float x, y, z; };

struct Vertex __attribute__((aligned(16))) vertices[2] = {
       {  0.0f,   0.0f, 0.0f},
       {480.0f, 272.0f, 0.0f},
};

#define BUF_WIDTH (512)
#define SCR_WIDTH (480)
#define SCR_HEIGHT (272)

int main(int argc, char* argv[]) {
   void* frameBufferPointer = getStaticVramBuffer(BUF_WIDTH, SCR_HEIGHT, GU_PSM_8888);
   void* depthBufferPointer = getStaticVramBuffer(BUF_WIDTH, SCR_HEIGHT, GU_PSM_4444);

   sceGuInit();

   sceGuStart(GU_DIRECT, list);
   {
      sceGuDrawBuffer(GU_PSM_8888, frameBufferPointer, BUF_WIDTH);
      sceGuDispBuffer(SCR_WIDTH, SCR_HEIGHT, frameBufferPointer, BUF_WIDTH);
      sceGuDepthBuffer(depthBufferPointer, BUF_WIDTH);
      sceGuOffset(2048 - (SCR_WIDTH / 2), 2048 - (SCR_HEIGHT / 2));
      sceGuViewport(2048, 2048, SCR_WIDTH, SCR_HEIGHT);
      sceGuDepthRange(65535, 0);
      sceGuScissor(0, 0, SCR_WIDTH, SCR_HEIGHT);
      sceGuEnable(GU_SCISSOR_TEST);
      sceGuFrontFace(GU_CW);
      sceGuShadeModel(GU_SMOOTH);
      sceGuDisable(GU_TEXTURE_2D);
   }
   sceGuFinish();
   sceGuSync(0, 0);

   sceGuDisplay(1);

   {
      sceGuStart(GU_DIRECT,list);
      {
         sceGuClearColor(0);
         sceGuClearDepth(0);
         sceGuClear(GU_COLOR_BUFFER_BIT | GU_DEPTH_BUFFER_BIT);

         sceGumMatrixMode(GU_PROJECTION); sceGumLoadIdentity();
         sceGumMatrixMode(GU_VIEW); sceGumLoadIdentity();
         sceGumMatrixMode(GU_MODEL); sceGumLoadIdentity();

         sceGuColor(0xFFFFFFFF);

         sceGumDrawArray(GU_LINES, GU_VERTEX_32BITF | GU_TRANSFORM_2D, 2, 0, vertices);
      }
      sceGuFinish();
      sceGuSync(0, 0);
   }

   // It will check the contents of the frameBuffer.
   emitMemoryBlock((void *)((unsigned int)frameBufferPointer | 0x04000000), SCR_WIDTH * SCR_HEIGHT * 4);
   
   sceGuTerm();
   sceKernelExitGame();
   return 0;
}
