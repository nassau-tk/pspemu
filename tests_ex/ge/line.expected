memory:0x00000000